jenkins/scripts/fetch-failed-request-logs.py

```diff
@@ -0,0 +1,200 @@
+# (or --yesterday) need to be provided on the cmd line.  A prometheus-style
+# prometheus_metrics.txt file is generated, and logs are downloaded into a logs/ subdirectory
+#
+
+# example usage - fetch-failed-request-logs.py --yesterday --server my.cachito.server.example.com
+#
+
+import argparse
+import datetime
+import json
+import os
+import sys
+from datetime import date, timedelta
+from textwrap import dedent
+
+import urllib3
+from icecream import ic
+
+# Cachito API Path.
+API_PATH = "api/v1"
+
+# Failed state
+STATE = "failed"
+
+# Disable all debug statements for prod.
+ic.enable()
+
+
+def _parse_cli():
+    parser = argparse.ArgumentParser(description="Fetch failed Cachito jobs in a given time range",
+                                     epilog="Either --date OR --yesterday is required.")
+    group = parser.add_mutually_exclusive_group(required=True)
+    group.add_argument(
+        "--date",
+        dest="date",
+        help="Date to examine for failed Cachito requests, in Year-Mon-Date format.  "
+        + "Example: 2021-11-16",
+    )
+    group.add_argument(
+        "--yesterday",
+        action="store_true",
+        help="Return failures from yesterday.  Causes script to ignore the --date argument",
+    )
+    parser.add_argument("--server", dest="cachitoserver", help="Cachito Server Hostname",
+                        required=True)
+    return parser.parse_args()
+
+
+def _is_same_date(item, request_date):
+    date_obj = datetime.datetime.strptime(item["updated"], "%Y-%m-%dT%H:%M:%S.%f").date()
+    if request_date == date_obj:
+        return 1
+
+
+def _fetch_cachito_jobs(cachitoserver, request_date, state):
+    # Cachito response for requests is reverse chronological order, so newest fist.
+
+    # API args
+    #
+    # -  page=int
+    # -  per_page=int
+    # -  state=string
+    # -  created_from=datetime
+    # -  created_to=datetime
+    #
+
+    # Cachito config currently has hard limit of 100.
+    per_page_requests = 100
+    page = 1
+    http = urllib3.PoolManager()
+    cachito_failed_requests = []
+    cachito_total_requests = []
+    api_url = "http://" + cachitoserver + "/" + API_PATH
+
+    # Initial URL.  The meta section of the response gives the 'next' URL with
+    # an updated page value, and 'None' when at the end of the available
+    # requests based on the query.  Using the 'next' URL to simplify the logic here.
+    request_url = (
+        api_url
+        + "/requests"
+        + "?per_page="
+        + str(per_page_requests)
+        + "&page="
+        + str(page)
+        + "&created_from="
+        + str(request_date)
+        + "&created_to="
+        + str(request_date)
+    )
+    while True:
+        r = http.request("GET", ic(request_url))
+        json_response = json.loads(r.data)
+        ic(json_response)
+
+        cachito_total_requests.extend(
+            [item for item in json_response["items"] if _is_same_date(item, request_date)]
```

Comment:
No, we should not.  The "created" date is the one used to prevent this exact problem, as "updated" moves.

Reply:
Ok, let me provide an example:
We need to filter all requests completed or failed yesterday.
R1: 2021-11-13T23:59 - 2021-11-14T00:01
R2: 2021-11-14T12:00 - 2021-11-14T12:30
R3: 2021-11-14T23:59 - 2021-11-15T00:01
Expected result: R1, R2
Actual result: R2
Because first filter there is by created date and we will fetch only R2 and R3.

Reply:
So just to clarify, the Actual Result above should be "R2, R3"?    

In this case, when the script was run on the 14th (yesterday), R1 would have been one of the last items reported.  I'm less concerned about the corner case of requests falling over the date transition, as long as they're reported on one side or the other of the transition, and not both. Using the created date, which never changes should assure us that a request (and its failure, if there is one) is reported only once on one date.

Reply:
We lose some results though.
I expect to see R3 in `--date=2021-11-15`, but it will be missed. In practice it could be a bunch of requests.

I would suggest to use `/request-metrics` here instead of `/requests`:
```http
GET /api/v1/request-metrics?finished_from=2021-11-14&finished_to=2021-11-14
```
The only one concern here, the endpoint is not stable enough.

Reply:
> We lose some results though. I expect to see R3 in `--date=2021-11-15`, but it will be missed. In practice it could be a bunch of requests.

Its not lost.  R3's status (failed or successful) was reported on the run for 2021-11-14.

Reply:
So you changed the _is_same_date filter, and now it check `created` but not `updated`, right?
If so, is there any sense to keep the filter? Given that the created_from/created_to filter does the job.

Reply:
Correct, it checks created.   As far as the filter, yes, I can remove it.  Hadn't crossed my mind.